File: frontend/src/services/issueLinkService.ts
import api from '@/lib/api'

export interface IssueLink {
    id: string
    sourceId: string
    targetId: string
    type: 'blocks' | 'is_blocked_by' | 'relates_to' | 'duplicates'
    source?: any
    target?: any
    createdAt: string
}

export const issueLinkService = {
    create: async (data: { sourceId: string; targetId: string; type: string }) => {
        const response = await api.post('/api/links', data)
        return response.data
    },

    delete: async (linkId: string) => {
        await api.delete(`/api/links/${linkId}`)
    },

    getByIssue: async (issueId: string) => {
        const response = await api.get(`/api/links/issue/${issueId}`)
        return response.data as IssueLink[]
    }
}

File: backend/internal/services/issue_link_service.go
package services

import (
	"errors"

	"github.com/braviz/jira-clone/internal/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type IssueLinkService struct {
	db *gorm.DB
}

func NewIssueLinkService(db *gorm.DB) *IssueLinkService {
	return &IssueLinkService{db: db}
}

func (s *IssueLinkService) CreateLink(sourceID, targetID uuid.UUID, linkType string) (*models.IssueLink, error) {
	if sourceID == targetID {
		return nil, errors.New("cannot link issue to itself")
	}

	// Check if link already exists
	var count int64
	s.db.Model(&models.IssueLink{}).
		Where("source_id = ? AND target_id = ? AND type = ?", sourceID, targetID, linkType).
		Count(&count)

	if count > 0 {
		return nil, errors.New("link already exists")
	}

	link := &models.IssueLink{
		SourceID: sourceID,
		TargetID: targetID,
		Type:     linkType,
	}

	if err := s.db.Create(link).Error; err != nil {
		return nil, err
	}

	// Preload relationships for return
	if err := s.db.Preload("Source").Preload("Target").First(link, link.ID).Error; err != nil {
		return nil, err
	}

	return link, nil
}

func (s *IssueLinkService) DeleteLink(linkID uuid.UUID) error {
	return s.db.Delete(&models.IssueLink{}, linkID).Error
}

func (s *IssueLinkService) GetIssueLinks(issueID uuid.UUID) ([]models.IssueLink, error) {
	var links []models.IssueLink
	// Get both outgoing (Source = issueID) and incoming (Target = issueID) links
	err := s.db.Preload("Source").Preload("Target").
		Where("source_id = ? OR target_id = ?", issueID, issueID).
		Find(&links).Error
	return links, err
}
